### ISSUE TITLE
[Suggestion] Add an "edges"-like/cursor result that returns the cursor of each row

### ISSUE THREAD

**Issue by cerinoligutom** [olavim/objection-cursor#17]
Something that returns a shape like the one below. 

```ts
interface ICursorPaginationResult<M> {
  nodes: ICursorResult<M>[];  // this one
  results: M[]
  pageInfo: IPageInfo;
}

interface ICursorResult<M> {
  cursor: string;
  data: M;
}

interface IPageInfo {
  // Current page info returned by objection-cursor
}
```

Our use case is we use GraphQL with connection types and below is our format of cursor based pagination results from the persistence layer.

```ts
interface ICursorPaginationResult<M> {
  results: ICursorResult<M>[];
  pageInfo: IPageInfo;
  totalCount: number;
  remaining: number;
}

interface ICursorResult<M> {
  cursor: string;
  data: M;
}

interface IPageInfo {
  startCursor: string | null;
  hasNextPage: boolean;
  hasPreviousPage: boolean;
  endCursor: string | null;
  // ... others
}
```

We're interested in getting the cursor of each resulting row with your plugin, the seamless serialization process + handling nullable fields using coalescing on top of ObjectionJS, thus the introduction of `ICursorResult`.

Probably to make it backwards compatible, add a new property called `nodes` that only gets returned (that has the shape of `ICursorResult`) if specified in the mixin config? I think a lot of GraphQL users that uses ObjectionJS w/ this plugin would benefit a lot from it.

**Comment by olavim** [comment]
Sounds good to me. I think I'd prefer calling them edges, as it seems more common terminology in the context of cursor pagination.

```ts
interface ICursorPaginationResult<M> {
  edges: IEdge<M>[];
  // ...
}

interface IEdge<M> {
  cursor: string;
  node: M;
}
```

I can deprecate `results` with this, since `edges` is simply more powerful (initially `edges` is a mixin option, disabled by default).

**Comment by cerinoligutom** [comment]
Great! Looking forward to it! When do you think can we have it out? 😁 

Also, I think it only make sense as an `edge` once that's graphed or in the interface layer (GraphQL) since edges are supposed to optionally contain "relationships" in a graph. But when in the context of persistence layer I think it is rather vague imo so I'd like to look at the row/model being returned as a plain cursor row result since it may not be a graph-like (with the help of `.eager()`) result.

What do you think?

**Comment by olavim** [comment]
> I think it only make sense as an edge once that's graphed or in the interface layer

I take it you mean that the plain `results` array shouldn't be deprecated since it's more convenient when you don't need `edges`. Which makes sense to me.

> I'd like to look at the row/model being returned as a plain cursor row result since it may not be a graph-like (with the help of .eager()) result.

I'm not sure I follow. Whatever Objection returns is simply put into the `results` array, or in the case of edges, each result is wrapped inside an edge. The plugin is agnostic about the details and relationships they may contain.

**Comment by olavim** [comment]
> When do you think can we have it out?

I have a big refactoring almost done, and I'll start working on this once I get that out of the way. Hopefully I can get an alpha release out in a few days and use you as a guinea pig :wink:

**Comment by cerinoligutom** [comment]
> > I think it only make sense as an edge once that's graphed or in the interface layer
> 
> I take it you mean that the plain `results` array shouldn't be deprecated since it's more convenient when you don't need `edges`. Which makes sense to me.

No not really, just the name (`edges`) of the property 🙂 A simple `.map()` will do the trick.

> > I'd like to look at the row/model being returned as a plain cursor row result since it may not be a graph-like (with the help of .eager()) result.
> 
> I'm not sure I follow. Whatever Objection returns is simply put into the `results` array, or in the case of edges, each result is wrapped inside an edge. The plugin is agnostic about the details and relationships they may contain.

Now that I think about it that way, as a plugin, you're right. Was occupied with the relationships it may contain, scratch that.

> > When do you think can we have it out?
> 
> I have a big refactoring almost done, and I'll start working on this once I get that out of the way. Hopefully I can get an alpha release in a few days and use you as a guinea pig 😉

Neat! Will definitely test it out right away, let me know by then 😁

**Comment by olavim** [comment]
> No not really, just the name (edges) of the property

Ah, I see. Do you have a better idea for the name? `nodes` is the most accurate name I can think of, but it doesn't really communicate how it differs from `results`, and definitely doesn't remove the vagueness of its purpose. This point is of course moot once `results` property is removed, so might as well live with it for now.

**Comment by olavim** [comment]
Things went a bit more smoothly than I anticipated... can you check if version `1.2.0-alpha.0` works for you.

The return value will have a `nodes` property, which is an array:

```js
[
  {
    data: // row data
    cursor: // opaque string
  }
]
```

You need to set `nodes: true` in the options like this:

```js
{
  // results: true,
  nodes: true,
  pageInfo: {
    // ...
  }
}
```

**Comment by cerinoligutom** [comment]
> Things went a bit more smoothly than I anticipated... can you check if version `1.2.0-alpha.0` works for you.

Will test!

**Comment by cerinoligutom** [comment]
@olavim Works great! However, if I set `results` to `false` in the mixin config it throws an error 🙂 

![image](https://user-images.githubusercontent.com/6721822/75994542-172e3b80-5f36-11ea-87a0-c82f3746926e.png)

---

I should've came back here earlier. Was stuck trying to get code completion in TypeScript since the method is not being recognized (Related to #1 and #12). 

Thought I could get it by [extending the query builder](https://vincit.github.io/objection.js/recipes/custom-query-builder.html#extending-the-query-builder-in-typescript) but nope lol. The `cursorPage()` method is `undefined` on runtime so when used, it throws that `nextCursorPage()`/`previousCursorPage()` is not a function...

![image](https://user-images.githubusercontent.com/6721822/75995053-d125a780-5f36-11ea-8c71-c58094ed12c0.png)

That `any` ticks me so much. Doing type assertion everytime seems a hassle as well but is better than `any`. Ugh.

![image](https://user-images.githubusercontent.com/6721822/75995116-e995c200-5f36-11ea-81ea-aa15294b79b5.png)

![image](https://user-images.githubusercontent.com/6721822/75995160-fb776500-5f36-11ea-9497-a4275d6d2afc.png)

**Comment by olavim** [comment]
The error is fixed in `1.2.0-alpha.1`.

As for your TS endeavors, you can try and see if cherry picking some ideas from the [ts-rewrite branch](https://github.com/olavim/objection-cursor/blob/ts-rewrite/src/mixin.ts) is enough for you (see the value/type of return values, also `cursorPage` method in `src/query-builder.ts`).

**Comment by cerinoligutom** [comment]
I got it working. Been avoiding this approach as it's not the best, but it works lol.

Here's my query:

![image](https://user-images.githubusercontent.com/6721822/76004338-49469a00-5f44-11ea-8925-038b01586b56.png)

As you can see, there's no squiggly line for it. I just simply augmented the involved modules.

![image](https://user-images.githubusercontent.com/6721822/76006429-6cbf1400-5f47-11ea-8a01-0d62cdd8cf37.png)

---

What I want to ultimately do is override/replace your `previousCursorPage()` and `nextCursorPage()` alias methods on runtime to apply a factory function that maps your `cursorPage()` result to the data structure I need.

![image](https://user-images.githubusercontent.com/6721822/76006385-5b760780-5f47-11ea-9edd-d48bdd0d5e63.png)

Do you have an idea on how to pull this off?

**Comment by cerinoligutom** [comment]
> The error is fixed in `1.2.0-alpha.1`.
> 

Works now, awesome 👍

**Comment by olavim** [comment]
> Do you have an idea on how to pull this off?

You can do modifications on the returned data in the [`runAfter` method](https://vincit.github.io/objection.js/api/query-builder/other-methods.html#runafter). If you want to modify the result every time, I think your best bet is to override the `cursorPage` method and do something like this:

```js
export class BaseModel extends EnhancedModel {
  static get QueryBuilder() {
    return class extends EnhancedModel.QueryBuilder {
      /* No need to override previousCursorPage and nextCursorPage separately
       * since they just call this method with different arguments
       */
      cursorPage(...args) {
        return super
          .cursorPage(...args)
          .runAfter(result => {
            // do conversion on result here

            return result;
          });
      }
    }
  }
}
```

> Works now, awesome 👍

Great! Let me know if you have problems.

**Comment by cerinoligutom** [comment]
> > Do you have an idea on how to pull this off?
> 
> You can do modifications on the returned data in the [`runAfter` method](https://vincit.github.io/objection.js/api/query-builder/other-methods.html#runafter). If you want to modify the result every time, I think your best bet is to override the `cursorPage` method and do something like this:
> 
> ```js
> export class BaseModel extends EnhancedModel {
>   static get QueryBuilder() {
>     return class extends EnhancedModel.QueryBuilder {
>       /* No need to override previousCursorPage and nextCursorPage separately
>        * since they just call this method with different arguments
>        */
>       cursorPage(...args) {
>         return super
>           .cursorPage(...args)
>           .runAfter(result => {
>             // do conversion on result here
> 
>             return result;
>           });
>       }
>     }
>   }
> }
> ```
> 
> > Works now, awesome 👍
> 
> Great! Let me know if you have problems.

Tried your code but it's throwing all sorts of errors, mostly TypeScript complaining. I figured this (screenshot below) would be equivalent but now I get a runtime error. Not quite sure what's going on under the hood.

![image](https://user-images.githubusercontent.com/6721822/76011617-d6432080-5f4f-11ea-8a03-6eb1ce7ab889.png)

```
[api] `QueryBuilder#mergeContext` method is deprecated. The `context` method now behaves just like `mergeContext`. `mergeContext` method will be removed in 3.0
[api] (node:18956) UnhandledPromiseRejectionWarning: Error: undefined passed as argument #2 for 'where' operation. Call skipUndefined() method to ignore the undefined values.
[api]     at shouldBeAdded (C:\Users\ZeferiniX\Documents\Source Codes\ZeferiniX\GraphQL-Starter\node_modules\objection\lib\queryBuilder\operations\ObjectionToKnexConvertingOperation.js:44:15)
[api]     at KnexOperation.onAdd (C:\Users\ZeferiniX\Documents\Source Codes\ZeferiniX\GraphQL-Starter\node_modules\objection\lib\queryBuilder\operations\ObjectionToKnexConvertingOperation.js:24:12)
[api]     at MyQueryBuilder.callOperationMethod (C:\Users\ZeferiniX\Documents\Source Codes\ZeferiniX\GraphQL-Starter\node_modules\objection\lib\queryBuilder\QueryBuilderOperationSupport.js:350:33)
[api]     at MyQueryBuilder.addOperationUsingMethod (C:\Users\ZeferiniX\Documents\Source Codes\ZeferiniX\GraphQL-Starter\node_modules\objection\lib\queryBuilder\QueryBuilderOperationSupport.js:381:28)
[api]     at MyQueryBuilder.addOperation (C:\Users\ZeferiniX\Documents\Source Codes\ZeferiniX\GraphQL-Starter\node_modules\objection\lib\queryBuilder\QueryBuilderOperationSupport.js:372:22)
[api]     at MyQueryBuilder.where (C:\Users\ZeferiniX\Documents\Source Codes\ZeferiniX\GraphQL-Starter\node_modules\objection\lib\queryBuilder\QueryBuilderBase.js:156:17)
[api]     at MyQueryBuilder.<anonymous> (C:\Users\ZeferiniX\Documents\Source Codes\ZeferiniX\GraphQL-Starter\node_modules\objection-cursor\lib\operations\CursorPageOperation.js:196:8)
[api]     at convertQueryBuilderFunction (C:\Users\ZeferiniX\Documents\Source Codes\ZeferiniX\GraphQL-Starter\node_modules\objection\lib\queryBuilder\operations\ObjectionToKnexConvertingOperation.js:135:8)
[api]     at Builder.convertedKnexArgumentFunction (C:\Users\ZeferiniX\Documents\Source Codes\ZeferiniX\GraphQL-Starter\node_modules\objection\lib\queryBuilder\operations\ObjectionToKnexConvertingOperation.js:122:7)
[api]     at Formatter.compileCallback (C:\Users\ZeferiniX\Documents\Source Codes\ZeferiniX\GraphQL-Starter\node_modules\knex\lib\formatter.js:221:14)
[api]     at Formatter.rawOrFn (C:\Users\ZeferiniX\Documents\Source Codes\ZeferiniX\GraphQL-Starter\node_modules\knex\lib\formatter.js:167:36)
[api]     at QueryCompiler_PG.whereWrapped (C:\Users\ZeferiniX\Documents\Source Codes\ZeferiniX\GraphQL-Starter\node_modules\knex\lib\query\compiler.js:686:32)
[api]     at QueryCompiler_PG.where (C:\Users\ZeferiniX\Documents\Source Codes\ZeferiniX\GraphQL-Starter\node_modules\knex\lib\query\compiler.js:386:34)
[api]     at QueryCompiler_PG.toSQL (C:\Users\ZeferiniX\Documents\Source Codes\ZeferiniX\GraphQL-Starter\node_modules\knex\lib\query\compiler.js:64:29)
[api]     at Formatter.compileCallback (C:\Users\ZeferiniX\Documents\Source Codes\ZeferiniX\GraphQL-Starter\node_modules\knex\lib\formatter.js:228:21)
[api]     at Formatter.rawOrFn (C:\Users\ZeferiniX\Documents\Source Codes\ZeferiniX\GraphQL-Starter\node_modules\knex\lib\formatter.js:167:36)
[api] From previous event:
[api]     at processImmediate (internal/timers.js:456:21)
[api] From previous event:
[api]     at Runner.ensureConnection (C:\Users\ZeferiniX\Documents\Source Codes\ZeferiniX\GraphQL-Starter\node_modules\knex\lib\runner.js:259:8)
[api]     at Runner.run (C:\Users\ZeferiniX\Documents\Source Codes\ZeferiniX\GraphQL-Starter\node_modules\knex\lib\runner.js:27:12)
[api]     at Builder.Target.then (C:\Users\ZeferiniX\Documents\Source Codes\ZeferiniX\GraphQL-Starter\node_modules\knex\lib\interface.js:15:43)
[api] (node:18956) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` 
(see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 2)
[api] (node:18956) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

**Comment by olavim** [comment]
Difficult to say. For me it works just fine. Is your prototype-modifying code still in use? Could it be you're ordering the results by some nullable column without coalescing?

If you can't figure it out, you can open objection-cursor source in your node_modules and add a console log before [this line](https://github.com/olavim/objection-cursor/blob/master/lib/operations/CursorPageOperation.js#L39):
```js
// ...
console.log(this.keyset);
whereMore(builder, this.keyset);
// ...
```
and after [this line](https://github.com/olavim/objection-cursor/blob/master/lib/operations/CursorPageOperation.js#L47):
```js
// ...
onAfter(builder, data) {
  console.log(data);
  // ...
```
and show me the output.

**Comment by cerinoligutom** [comment]
> Difficult to say. For me it works just fine. Is your prototype-modifying code still in use? Could it be you're ordering the results by some nullable column without coalescing?

No, I removed the prototype code. The fields are non-nullable.

> If you can't figure it out, you can open objection-cursor source in your node_modules and add a console log before [this line](https://github.com/olavim/objection-cursor/blob/master/lib/operations/CursorPageOperation.js#L39):
> 
> ```js
> // ...
> console.log(this.keyset);
> whereMore(builder, this.keyset);
> // ...
> ```

![image](https://user-images.githubusercontent.com/6721822/76015265-efe76680-5f55-11ea-90fb-a8864f5822ae.png)


> and after [this line](https://github.com/olavim/objection-cursor/blob/master/lib/operations/CursorPageOperation.js#L47):
> 
> ```js
> // ...
> onAfter(builder, data) {
>   console.log(data);
>   // ...
> ```
> 

Didn't show the console log so I'm betting it didn't reach this point.

**Comment by olavim** [comment]
Please log `comparisons` after [this line](https://github.com/olavim/objection-cursor/blob/master/lib/operations/CursorPageOperation.js#L169)

**Comment by cerinoligutom** [comment]
![image](https://user-images.githubusercontent.com/6721822/76016239-7486b480-5f57-11ea-9245-a8c97bdd9408.png)

**Comment by cerinoligutom** [comment]
Here are my table results.
![image](https://user-images.githubusercontent.com/6721822/76016407-b7e12300-5f57-11ea-809d-4d2a8ed9fbe6.png)

Abel Abshire is in the middle as that cursor is after the first `cursorPage()` result.

**Comment by olavim** [comment]
Does your code somehow call `orderBy` twice for `firstName` and `lastName`? In any case, looks like a bug in the plugin for now. I'll try to figure it out. Could you show your query (code)?

**Comment by cerinoligutom** [comment]
> Does your code somehow call `orderBy` twice for `firstName` and `lastName`? In any case, looks like a bug in the plugin for now. I'll try to figure it out. Could you show your query (code)?

Yes it does.

```ts
const result = await User.query()
    .select()
    .orderBy('firstName')
    .orderBy('lastName')
    .limit(5)
    .nextCursorPage('KGpzb24pIkFiZWwi.KGpzb24pIkFic2hpcmUi');
```

**Comment by olavim** [comment]
I was able to reproduce your problem, but the way I did it is definitely not a bug in the plugin (although errors could be improved).

Here's my model (made it similar to yours):
```js
Model.knex(knex);

const EnhancedModel = compose([
  cursorPagination({
    nodes: true,
    results: false,
    pageInfo: {
      total: true,
      remaining: true,
      hasNext: true,
      hasPrevious: true
    }
  })
])(Model);

class EnhancedMovie extends EnhancedModel {
  static get tableName() {
    return 'movies';
  }

  static get QueryBuilder() {
    return class extends EnhancedModel.QueryBuilder {
      cursorPage(...args) {
        return super
          .cursorPage(...args)
          .runAfter(res => {
            return {res};
          });
      }
    }
  }
}
```

Get cursor:
```js
const {res} = await EnhancedMovie.query()
  .orderBy('author')
  .orderBy('id')
  .nextCursorPage();
// next => KGpzb24pImF1dGhvci00Ig.KGpzb24pMjA.KGpzb24pImF1dGhvci00Ig.KGpzb24pMjA
```

Use cursor, but add another `orderBy` statement:
```js
const {res} = await EnhancedMovie.query()
  .orderBy('author')
  .orderBy('id')
  .orderBy('title')
  .nextCursorPage('KGpzb24pImF1dGhvci00Ig.KGpzb24pMjA.KGpzb24pImF1dGhvci00Ig.KGpzb24pMjA');
// The cursor used here only supports two orderBy statements, and so this throws an error
```

```
     Error: undefined passed as argument #2 for 'where' operation. Call skipUndefined() method to ignore the undefined values.
      at shouldBeAdded (node_modules/objection/lib/queryBuilder/operations/ObjectionToKnexConvertingOperation.js:44:15)
      at KnexOperation.onAdd (node_modules/objection/lib/queryBuilder/operations/ObjectionToKnexConvertingOperation.js:24:12)
      at Object.callOperationMethod (node_modules/objection/lib/queryBuilder/QueryBuilderOperationSupport.js:350:33)
      at Object.addOperationUsingMethod (node_modules/objection/lib/queryBuilder/QueryBuilderOperationSupport.js:381:28)
      at Object.addOperation (node_modules/objection/lib/queryBuilder/QueryBuilderOperationSupport.js:372:22)
      at Object.where (node_modules/objection/lib/queryBuilder/QueryBuilderBase.js:156:17)
      at Object.<anonymous> (lib/operations/CursorPageOperation.js:196:8)
```

In your case this would happen if you had something like this:
```js
const result = await User.query()
    .select()
    .orderBy('firstName')
    .orderBy('lastName')
    .orderBy('firstName')
    .orderBy('lastName')
    .limit(5)
    .nextCursorPage();
```

Then you got the cursor, and removed the duplicate `orderBy`s.

---

If this isn't the case, then please make a repository with code that I can run and get the error.

**Comment by cerinoligutom** [comment]
I noticed there are 4 segments in the cursor now. Was using the cursor generated on `alpha.0` when I tried to get results on `alpha.1`, could that be the culprit? That aside, it's working with a fresh cursor now. 

Now I noticed a different problem. 

This is the query (same)
![image](https://user-images.githubusercontent.com/6721822/76078366-7a26dd80-5fdd-11ea-96a5-302f97775421.png)

and here's how the BaseModel looks now (based on your code but in TS)

![image](https://user-images.githubusercontent.com/6721822/76078387-8743cc80-5fdd-11ea-86b9-6090cd38c905.png)

If I set the static `QueryBuilder` (regardless of what's inside), the `orderBy()` builder gets ignored by your plugin so the results are as good as not having the `ORDER BY` statements. Could this be TS-specific?

**Comment by cerinoligutom** [comment]
You can clone [this repo](https://github.com/cerino-ligutom/GraphQL-Starter/tree/objection-cursor-testing) then switch to **objection-cursor-testing** branch. 

To get it up and running for this particular branch:
```bash
$ npm install
$ npm start
```

Ignore the docker stuffs as I only use the database temporarily for this branch since the node alpine image does not have git so I cannot pull the github repo inside the container.

Modify the `knexfile.ts` as per your database configuration. I have some npm scripts for the knex migration if that helps set up the table I'm testing it against.

[The query is at @ `src\app.ts`](https://github.com/cerino-ligutom/GraphQL-Starter/blob/5f5921944102002821b3a7af33f8349286a8616b/src/app.ts#L25) 

Experiment with the `BaseModel` @ `src\db\models\common\BaseModel.ts`

**Comment by olavim** [comment]
Thanks for the repo. Managed to fix it already 😅 can you try `1.2.0-alpha.2` for me

**Comment by cerinoligutom** [comment]
> Thanks for the repo. Managed to fix it already 😅 can you try `1.2.0-alpha.2` for me

haha, logical error somewhere? 😁 Seems to be working now, my QueryBuilder isn't acting up anymore as well. Great!
